targeted/2xml4xxe.js

```diff
@@ -0,0 +1,248 @@
+// A Script to convert Request body in normal form or JSON to XML
+// and set Content type to application/json.
+// it uses manual request editor where you can edit the convered request 
+// it is not an automated tool in finding XXE 
+// it may be helpful in finding XXE or other vulnerabilities.
+// this script is intended to  act as an assistant
+// you can  add anything like [!ENTITY] to test in detail
+// released under the Apache v2.0 licence.
+// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+// Author : @haseebeqx (GitHub, Twitter)
+// tested on: ZAP 2.4.1
+// note : due to a bug in mozilla Rhino  bdy.charAt(j) had to be replaced with String.fromCharCode(bdy.charAt(j)). so in future (or in Nashorn) you may have to change it back
+// rule1: pure JSON , no CODE
+// rule2: correct body (make edits only after conversion)
+
+function invokeWith(msg) {
+	body = '<?xml version="1.0" encoding="UTF-8"?>\n';
+	reqb = msg.getRequestBody().toString(); 
+	reqh = msg.getRequestHeader().getURI().toString();
+	if(isJson(reqb)){
+		body += jsonToXML(reqb);
+	}
+	else if(ismultipart(msg.getRequestHeader())){
+		js = multiToJson(msg);
+		body += jsonToXML(js);
+	}
+	else{
+		js = bodyToJson(reqb);
+		body += jsonToXML(js);
+	}
+	msg.setRequestBody(body);
+	header = msg.getRequestHeader();
+	header.setHeader(org.parosproxy.paros.network.HttpRequestHeader.CONTENT_TYPE,"application/xml");
+	header.setHeader(org.parosproxy.paros.network.HttpRequestHeader.CONTENT_LENGTH,body.length);
+	msg.setRequestHeader(header);
+	ext = new org.parosproxy.paros.extension.history.ExtensionHistory;
```

Comment:
Ok but all I want to do was manual request editor. I am at work. currently
all I have is a basic phone. I can not edit it :(

On Tue, Aug 25, 2015, 2:09 PM thc202 <notifications@github.com> wrote:

> In targeted/2xml4xxe.js
> <https://github.com/zaproxy/community-scripts/pull/18#discussion_r37844197>
> :
>
> > +		body += jsonToXML(reqb);
> > +	}
> > +	else if(ismultipart(msg.getRequestHeader())){
> > +		js = multiToJson(msg);
> > +		body += jsonToXML(js);
> > +	}
> > +	else{
> > +		js = bodyToJson(reqb);
> > +		body += jsonToXML(js);
> > +	}
> > +	msg.setRequestBody(body);
> > +	header = msg.getRequestHeader();
> > +	header.setHeader(org.parosproxy.paros.network.HttpRequestHeader.CONTENT_TYPE,"application/xml");
> > +	header.setHeader(org.parosproxy.paros.network.HttpRequestHeader.CONTENT_LENGTH,body.length);
> > +	msg.setRequestHeader(header);
> > +	ext = new org.parosproxy.paros.extension.history.ExtensionHistory;
>
> Use
> org.parosproxy.paros.control.Control.getSingleton().getExtensionLoader().getExtension(
> org.parosproxy.paros.extension.history.ExtensionHistory.NAME) instead, to
> obtain the existing ExtensionHistory.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zaproxy/community-scripts/pull/18/files#r37844197>.
>